service-impl/user-service-impl/ldap-user-service/src/main/java/uk/gov/gchq/palisade/user/service/impl/AbstractLdapUserService.java

```diff
@@ -0,0 +1,336 @@
+            userCache = CacheBuilder.newBuilder()
+                    .maximumSize(10000)
+                    .expireAfterWrite(AbstractLdapUserService.cacheTtlHours, TimeUnit.HOURS)
+                    .build();
+        }
+    }
+
+    protected Map<String, Object> getAttributes(final UserId userId) throws NamingException {
+        final Map<String, Object> attributes = new HashMap<>();
+        final String[] requestAttrs = getAttributeNames();
+        if (null != requestAttrs && requestAttrs.length > 0) {
+            final Attributes userAttrs = context.getAttributes(escapeLdapSearchFilter(userId.getId()), requestAttrs);
+            if (null != userAttrs) {
+                for (final String requestAttr : requestAttrs) {
+                    final Attribute attribute = userAttrs.get(requestAttr);
+                    if (null != attribute) {
+                        final NamingEnumeration<?> all = attribute.getAll();
+                        if (all.hasMore()) {
+                            attributes.put(requestAttr, all.next());
+                        }
+                    }
+                }
+            }
+        }
+        return attributes;
+    }
+
+    /**
+     * Performs a basic search on LDAP using the userId.
+     *
+     * @param userId          the userId to search for
+     * @param name            the name of the context to search
+     * @param attrIdForUserId the attribute ID that is associated with the userId
+     * @param attrs           the attributes to fetch from the LDAP search.
+     * @return the attribute values
+     * @throws NamingException if a naming exception is encountered
+     */
+    protected Set<Object> basicSearch(final UserId userId,
+                                      final String name, final String attrIdForUserId,
+                                      final String... attrs) throws NamingException {
+        final NamingEnumeration<SearchResult> attrResults = context.search(
+                name,
+                new BasicAttributes(attrIdForUserId, escapeLdapSearchFilter(userId.getId())),
+                attrs
+        );
+
+        final Set<Object> results = new HashSet<>();
+        while (attrResults.hasMore()) {
+            final SearchResult result = attrResults.next();
+            final Attributes resultAttrs = result.getAttributes();
+            if (null != resultAttrs) {
+                final NamingEnumeration<? extends Attribute> all = resultAttrs.getAll();
+                if (null != all) {
+                    while (all.hasMore()) {
+                        final Attribute next = all.next();
+                        final Object nextValue = next.get();
+                        if (null != nextValue) {
+                            results.add(nextValue);
+                        }
+                    }
+                }
+            }
+        }
+        return results;
+    }
+
+    protected String escapeLdapSearchFilter(final String input) {
```

Comment:
Is this intended to be overridden?

Reply:
Yes - I set it to protected to allow users to override it if they realised they wanted to escape other things as well.